File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.Scanner;

public class Main {
    static final int maxN = 24;
    static int[] arr = new int[maxN];
    static int N, S, cnt = 0;

     static void go(int n, int sum) {
        if (n >= N) {
            if (sum == S) cnt++;
            return;
        }
        
        go(n + 1, sum);
        go(n + 1, sum + arr[n]);
    }
    
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        S = sc.nextInt();
        
        for (int i = 0; i < N; i++) {
            arr[i] = sc.nextInt();
        }
        
        go(0, 0);
        
        if (S == 0) cnt--;
        System.out.println(cnt);
        
        sc.close();
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include<iostream>
#include<vector>
#include<cstring>
#include <cstdlib> 
using namespace std;

const int maxN = 24;
int arr[maxN];

int N, S;
int cnt;

void go(int n, int sum) {
	if (n >= N) {
		if (sum == S) cnt++;
		return;
	}
	go(n + 1, sum);
	go(n + 1, sum + arr[n]);
}


int main() {
	cin >> N >> S;
	for (int i = 0; i < N; i++) {
		cin >> arr[i];
	}
	vector<int> vec;
	go(0, 0);
	if (S == 0) cnt--;
	cout << cnt;

	return 0;
}